File: android/src/main/java/com/valentingrigorean/arcgis_maps_flutter/map/BaseGraphicController.java
package com.valentingrigorean.arcgis_maps_flutter.map;

import android.graphics.Color;

import com.esri.arcgisruntime.geometry.Geometry;
import com.esri.arcgisruntime.mapping.view.Graphic;
import com.esri.arcgisruntime.mapping.view.GraphicsOverlay;

public abstract class BaseGraphicController implements GraphicControllerSink {

    private SelectionPropertiesHandler selectionPropertiesHandler;
    private boolean consumeTapEvents;
    private Color selectedColor;


    protected abstract Graphic getGraphic();

    public void setSelectionPropertiesHandler(SelectionPropertiesHandler selectionPropertiesHandler) {
        this.selectionPropertiesHandler = selectionPropertiesHandler;
    }

    public void add(GraphicsOverlay graphicsOverlay) {
        graphicsOverlay.getGraphics().add(getGraphic());
    }

    public void remove(GraphicsOverlay graphicsOverlay) {
        graphicsOverlay.getGraphics().remove(getGraphic());
    }

    public boolean getVisible() {
        return getGraphic().isVisible();
    }

    public void setVisible(boolean visible) {
        getGraphic().setVisible(visible);
    }

    public void setZIndex(int zIndex) {
        getGraphic().setZIndex(zIndex);
    }

    public void setGeometry(Geometry geometry) {
        getGraphic().setGeometry(geometry);
    }

    public boolean isSelected() {
        return getGraphic().isSelected();
    }

    public void setSelected(boolean selected) {
        final Graphic graphic = getGraphic();
        if (selected && !graphic.isSelected()) {
            selectionPropertiesHandler.setSelected(graphic, selectedColor);
        } else if (graphic.isSelected()) {
            selectionPropertiesHandler.clearSelection(graphic);
        }
    }

    public void setConsumeTapEvents(boolean consumeTapEvents) {
        this.consumeTapEvents = consumeTapEvents;
    }

    public boolean canConsumeTapEvents() {
        return consumeTapEvents;
    }


    public void setConsumeTapEvent(boolean consumeTapEvent) {
        this.consumeTapEvents = consumeTapEvent;
    }

    public void setSelectedColor(Color selectedColor) {
        this.selectedColor = selectedColor;
    }
}

File: android/src/main/java/com/valentingrigorean/arcgis_maps_flutter/mapping/symbology/BaseGraphicController.kt
package com.valentingrigorean.arcgis_maps_flutter.mapping.symbology

import com.arcgismaps.Color
import com.arcgismaps.geometry.Geometry
import com.arcgismaps.mapping.view.Graphic
import com.arcgismaps.mapping.view.GraphicsOverlay
import com.valentingrigorean.arcgis_maps_flutter.convert.map.toSymbolVisibilityFilterOrNull
import com.valentingrigorean.arcgis_maps_flutter.convert.toArcgisColor
import com.valentingrigorean.arcgis_maps_flutter.map.SelectionPropertiesHandler
import com.valentingrigorean.arcgis_maps_flutter.map.SymbolVisibilityFilterController

abstract class BaseGraphicController(
    override var consumeTapEvents: Boolean = false,
    override var selectedColor: Color? = null,
) : GraphicControllerSink {
    private var selectionPropertiesHandler: SelectionPropertiesHandler? = null
    protected val graphic = Graphic()

    override var geometry: Geometry?
        get() = graphic.geometry
        set(value) {
            graphic.geometry = value
        }

    override var zIndex: Int
        get() = graphic.zIndex
        set(value) {
            graphic.zIndex = value
        }

    override var visible: Boolean
        get() = graphic.isVisible
        set(visible) {
            graphic.isVisible = visible
        }


    override var isSelected: Boolean
        get() = graphic.isSelected
        set(selected) {
            val graphic = graphic
            if (selected && !graphic.isSelected) {
                selectionPropertiesHandler?.setSelected(graphic, selectedColor)
            } else if (graphic.isSelected) {
                selectionPropertiesHandler?.clearSelection(graphic)
            }
        }


    fun setSelectionPropertiesHandler(selectionPropertiesHandler: SelectionPropertiesHandler?) {
        this.selectionPropertiesHandler = selectionPropertiesHandler
    }

    override fun add(graphicsOverlay: GraphicsOverlay) {
        graphicsOverlay.graphics.add(graphic)
    }

    override fun remove(graphicsOverlay: GraphicsOverlay) {
        graphicsOverlay.graphics.remove(graphic)
    }

    override fun interpretGraphicController(
        data: Map<*, *>,
        symbolVisibilityFilterController: SymbolVisibilityFilterController?,
    ) {
        val consumeTapEvents = data["consumeTapEvents"] as Boolean?
        if (consumeTapEvents != null) {
            this.consumeTapEvents = consumeTapEvents
        }
        val visible = data["visible"] as Boolean?
        if (visible != null) {
            val visibilityFilter = data["visibilityFilter"]?.toSymbolVisibilityFilterOrNull()
            if (symbolVisibilityFilterController != null && visibilityFilter != null) {
                symbolVisibilityFilterController!!.addGraphicsController(
                    this,
                    visibilityFilter,
                    visible
                )
            } else {
                this.visible = visible
            }
        }
        val zIndex = data["zIndex"] as Int?
        if (zIndex != null) {
            this.zIndex = zIndex
        }
        val selectedColor = (data["selectedColor"] as Int?)?.toArcgisColor()
        if (selectedColor != null) {
            this.selectedColor = selectedColor
        }
    }
}